public/app/features/dashboard-scene/variables/utils.ts

```diff
@@ -26,3 +30,219 @@ export const createDependencyEdges = (variables: Array<SceneVariable<SceneVariab
+  }
+
+  return usages;
+}
+
+export function transformUsagesToNetwork(usages: VariableUsageTree[]): UsagesToNetwork[] {
+  const results: UsagesToNetwork[] = [];
+
+  for (const usage of usages) {
+    const { variable, tree } = usage;
+    const result: UsagesToNetwork = {
+      variable,
+      nodes: [{ id: 'dashboard', label: 'dashboard' }],
+      edges: [],
+      showGraph: false,
+    };
+    results.push(traverseTree(result, { id: 'dashboard', value: tree }));
+  }
+
+  return results;
+}
+
+export const traverseTree = (usage: UsagesToNetwork, parent: { id: string; value: unknown }): UsagesToNetwork => {
+  const { id, value } = parent;
+  const { nodes, edges } = usage;
+
+  if (value && typeof value === 'string') {
+    const leafId = `${parent.id}-${value}`;
+    nodes.push({ id: leafId, label: value });
+    edges.push({ from: leafId, to: id });
+
+    return usage;
+  }
+
+  if (isRecord(value)) {
+    const keys = Object.keys(value);
+    for (const key of keys) {
+      const leafId = `${parent.id}-${key}`;
+      nodes.push({ id: leafId, label: key });
+      edges.push({ from: leafId, to: id });
+      usage = traverseTree(usage, { id: leafId, value: value[key] });
+    }
+
+    return usage;
+  }
+
+  return usage;
+};
+
+export const getVariableUsages = (variableId: string, usages: VariableUsageTree[]): number => {
+  const usage = usages.find((usage) => usage.variable.state.name === variableId);
+  if (!usage) {
+    return 0;
+  }
+
+  if (isRecord(usage.tree)) {
+    return countLeaves(usage.tree);
+  }
+
+  return 0;
+};
+
+const countLeaves = (object: object): number => {
+  const total = Object.values(object).reduce<number>((count, value) => {
+    if (typeof value === 'object') {
+      return count + countLeaves(value);
+    }
+
+    return count + 1;
+  }, 0);
+
+  return total;
+};
+
+export async function getUnknownsNetwork(
+  variables: Array<SceneVariable<SceneVariableState>>,
+  dashboard: Dashboard | null
+): Promise<UsagesToNetwork[]> {
+  return new Promise((resolve, reject) => {
+    // can be an expensive call so we avoid blocking the main thread
+    setTimeout(() => {
+      try {
+        const unknowns = createUnknownsNetwork(variables, dashboard);
+        resolve(transformUsagesToNetwork(unknowns));
+      } catch (e) {
+        reject(e);
+      }
+    }, 200);
+  });
+}
+
+function createUnknownsNetwork(
+  variables: Array<SceneVariable<SceneVariableState>>,
+  dashboard: Dashboard | null
+): VariableUsageTree[] {
+  if (!dashboard) {
+    return [];
+  }
+
+  let unknown: VariableUsageTree[] = [];
+  const unknownVariables = getUnknownVariableStrings(variables, dashboard);
+  for (const unknownVariable of unknownVariables) {
+    const props = getPropsWithVariable(unknownVariable, { key: 'model', value: dashboard }, {});
+    if (Object.keys(props).length) {
+      const variable = {
+        state: {
+          name: unknownVariable,
+        },
+      } as SceneVariable<SceneVariableState>;
```

Comment:
This seems pretty problematic to me - SceneVariable looks like a pretty complex class/object with a bunch of methods which are just being cast away here?

Reply:
Not ideal but this is only for missing/renamed variables section where we access name purely for display purposes. We can either leave it like this or duplicate `transformUsagesToNetwork` and `traverseTree` functions just for this use case to get that type safety. What do you think @joshhunt is the lesser evil here?

Reply:
> where we access name purely for display purposes

For now. Until later on someone makes a change to the page with the assumption that more properties are there and you ship it and it crashes in production. "cannot read property foo of undefined" is the most embarrasing report to get from users because it's entirely preventable, yet we're telling typescript to ignore it.

I think we should use the type system for what it is and not just lie to it and our future selves.

If using the correct type is too difficult here, maybe another approach is needed. Does this _have_ to be a SceneVariable if all you're using is `name`?

Reply:
@joshhunt managed to address type safety with minimal changes